proposals/p1885.md

```diff
@@ -668,36 +701,115 @@ for (i: i32 in v as CppIterate(Cpp.std.vector(i32))) { ... }
 These two options could allow interoperability with a compatible C++ type,
 either implicitly (template approach) or explicitly (adapter approach).
 
-#### Supporting Carbon `Iterate` in C++
+#### Iterating over Carbon types in C++
 
-Support for `Iterate` in C++ needs to provide `begin()` and `end()` methods or
-free functions, in addition to `*` and `++` operators.
+C++ needs to provide `begin()` and `end()` methods or free functions, in
+addition to `*` and `++` operators.
 
-`Iterate` does not expose a way to have a cursor or iterator to the past-last
-element. An option would be to have `end()` return a predefined invalid value,
-to satisfy iterator comparison when `Next()` returns an empty `Optional`.
+The `Iterate` interface does not expose a way to have a cursor nor an iterator
+to the past-last element. An option would be to have `end()` return a predefined
+invalid value, to satisfy iterator comparison when `Next()` returns an empty
+`Optional`.
 
-Below is an incomplete example of classes exposing those methods:
+Below is sample implementation to iterate on a Carbon `Iterate`. Note that we
+need to cache the last value, to allow for `*it` to be called, and to be called
+more than once.
 
 ```cpp
-template <typename T>
-class Cursor {
-  Cursor<T> FromCarbonCursor(...);
-  Cursor<T> Invalid();
+#include <cassert>
+#include <iostream>
+#include <optional>
+#include <string>
+
+namespace Carbon {
+
+template <typename C, typename T>
+class Iterable;
+
+template <typename Iterable>
```

Comment:
Right, I battled that a minute, intending to use `Iterable` as a kind of concept, rather than the actual `Iterate`. Reads better, but agree its a bit more confusing.